File: leetcode/53/MaxSubArray.java
import javax.swing.SortOrder;

class MaxSubArray {
    public static void main(String[] args) {
        Solution sl = new Solution();
        int result = sl.maxSubArray(new int[] { -2, 1, -3, 4, -1, 2, 1, -5, 4 });
        System.out.println(result);
    }
}

class Solution {
    // public int maxSubArray(int[] nums) {

    // int max = nums[0];
    // int sum = 0;
    // for(int i = 0; i < nums.length; i++)
    // {
    // sum = 0;
    // for(int j = i; j < nums.length; j++)
    // {
    // sum += nums[j];
    // if(sum > max)
    // {
    // max = sum;
    // }
    // }
    // }
    // return max;
    // }

    // public int maxSubArray(int[] nums) {
    // if(nums == null || nums.length == 0) return 0;
    // int len = nums.length;
    // int curMax = nums[0], res = nums[0];
    // for(int i = 1; i < len; i++){
    // int tmpMax = curMax;
    // curMax = Math.max(nums[i], tmpMax+nums[i]);
    // res = Math.max(curMax, res);
    // }
    // return res;

    // }

    public int maxSubArray(int[] nums) {
        int sum = nums[0];
        for (int i = 1; i < nums.length; i++) {
            if (nums[i - 1] > 0) {
                nums[i] += nums[i - 1];
            }
            sum = Math.max(nums[i], sum);
        }
        return sum;
    }
}

File: leetcode/53/MaxSubArray.js
var maxSubArray = function(nums) {
    let sum = nums[0];
    for (let i = 1; i < nums.length; i++) {
        console.log("nums["+i+"] = "+nums[i]);
        console.log("nums["+(i-1)+"] = "+nums[i-1]);

        if (nums[i - 1] > 0) {
            console.log("nums["+i+"] += nums["+(i - 1)+"] = "+(nums[i] + nums[i - 1]));
            nums[i] += nums[i - 1];
        }
        
        console.log(nums);
        console.log("before max nums["+i+"] = "+nums[i],"sum : " + sum);
        sum = Math.max(nums[i], sum);
        console.log("after max nums["+i+"] = "+nums[i],"sum : " + sum);
        console.log("========");
        // sum = nums[i] > sum ? nums[i] : sum
    }
    return sum;
};
console.log(maxSubArray([-2, 1, -3, 4, -1, 2, 1, -5, 4]));

/* 
nums[1] = 1
nums[0] = -2
[
  -2, 1, -3, 4, -1,
   2, 1, -5, 4
]
before max nums[1] = 1 sum : -2
after max nums[1] = 1 sum : 1
========
nums[2] = -3
nums[1] = 1
nums[2] += nums[1] = -2
[
  -2, 1, -2, 4, -1,
   2, 1, -5, 4
]
before max nums[2] = -2 sum : 1
after max nums[2] = -2 sum : 1
========
nums[3] = 4
nums[2] = -2
[
  -2, 1, -2, 4, -1,
   2, 1, -5, 4
]
before max nums[3] = 4 sum : 1
after max nums[3] = 4 sum : 4
========
nums[4] = -1
nums[3] = 4
nums[4] += nums[3] = 3
[
  -2, 1, -2, 4, 3,
   2, 1, -5, 4
]
before max nums[4] = 3 sum : 4
after max nums[4] = 3 sum : 4
========
nums[5] = 2
nums[4] = 3
nums[5] += nums[4] = 5
[
  -2, 1, -2, 4, 3,
   5, 1, -5, 4
]
before max nums[5] = 5 sum : 4
after max nums[5] = 5 sum : 5
========
nums[6] = 1
nums[5] = 5
nums[6] += nums[5] = 6
[
  -2, 1, -2, 4, 3,
   5, 6, -5, 4
]
before max nums[6] = 6 sum : 5
after max nums[6] = 6 sum : 6
========
nums[7] = -5
nums[6] = 6
nums[7] += nums[6] = 1
[
  -2, 1, -2, 4, 3,
   5, 6,  1, 4
]
before max nums[7] = 1 sum : 6
after max nums[7] = 1 sum : 6
========
nums[8] = 4
nums[7] = 1
nums[8] += nums[7] = 5
[
  -2, 1, -2, 4, 3,
   5, 6,  1, 5
]
before max nums[8] = 5 sum : 6
after max nums[8] = 5 sum : 6
========
6
*/